jetty-util/src/main/java/org/eclipse/jetty/util/OpaqueIndex.java

```diff
@@ -0,0 +1,62 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2020 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under
+// the terms of the Eclipse Public License 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0
+//
+// This Source Code may also be made available under the following
+// Secondary Licenses when the conditions for such availability set
+// forth in the Eclipse Public License, v. 2.0 are satisfied:
+// the Apache License v2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.util;
+
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+public interface OpaqueIndex
+{
+
+    boolean contains(String key);
+
+    class Builder
+    {
+        private Set<String> contents = new LinkedHashSet<>();
+        private boolean caseSensitive;
+
+        public Builder caseSensitive(boolean caseSensitive)
+        {
+            this.caseSensitive = caseSensitive;
+            return this;
+        }
+
+        public Builder with(String key)
+        {
+            contents.add(key);
+            return this;
+        }
+
+        public OpaqueIndex build()
+        {
+            Trie<Boolean> result;
+            if (caseSensitive)
+                result = new ArrayTernaryTrie<>(false, contents.size());
+            else
+                result = new ArrayTrie<>(contents.size());
```

Comment:
Indeed, this was a too quick'n'dirty hack. But since we know at immutable index creation what its contents are going to be, we could calculate exactly what its exact capacity should be, which was my plan.